mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala

```diff
@@ -31,20 +32,28 @@ private[ml] case class ParsedRFormula(label: ColumnRef, terms: Seq[Term]) {
    * of the special '.' term. Duplicate terms will be removed during resolution.
    */
   def resolve(schema: StructType): ResolvedRFormula = {
-    var includedTerms = Seq[String]()
+    val dotTerms = expandDot(schema)
+    var includedTerms = Seq[Seq[String]]()
     terms.foreach {
+      case term: ColumnRef =>
+        includedTerms :+= Seq(term.value)
+      case ColumnInteraction(terms) =>
+        includedTerms ++= expandInteraction(schema, terms)
       case Dot =>
-        includedTerms ++= simpleTypes(schema).filter(_ != label.value)
-      case ColumnRef(value) =>
-        includedTerms :+= value
+        includedTerms ++= dotTerms.map(Seq(_))
       case Deletion(term: Term) =>
         term match {
-          case ColumnRef(value) =>
-            includedTerms = includedTerms.filter(_ != value)
+          case inner: ColumnRef =>
+            includedTerms = includedTerms.filter(_ != Seq(inner.value))
+          case ColumnInteraction(terms) =>
+            val fromInteraction = expandInteraction(schema, terms).map(_.toSet)
+            includedTerms = includedTerms.filter(t => !fromInteraction.contains(t.toSet))
           case Dot =>
             // e.g. "- .", which removes all first-order terms
-            val fromSchema = simpleTypes(schema)
-            includedTerms = includedTerms.filter(fromSchema.contains(_))
+            includedTerms = includedTerms.filter {
+              case Seq(t) => !dotTerms.contains(t)
+              case _ => true
+            }
           case _: Deletion =>
             assert(false, "Deletion terms cannot be nested")
```

Comment:
* not part of this PR: `throw new RuntimeException(...)`
* also not part of this PR: Shall we move `hasIntercept` to `ResolvedRFormula`? It is a little strange to have two places to store the fully parsed formula.